### PR TITLE
supply royalty

### PR DESCRIPTION
**Pull request by jgeary**

### CODE REVIEW

File: src/ERC721Drop.sol [hunk 1]
```diff
@@ -1267,6 +1272,38 @@ contract ERC721Drop is
         emit MintFeePayout(zoraFee, ZORA_MINT_FEE_RECIPIENT, success);
     }
 
+    function _requireCanMintQuantity(uint256 quantity) internal view {
+        if (quantity + _totalMinted() > config.editionSize) {
+            revert Mint_SoldOut();
+        }
+    }
+
+    function _handleSupplyRoyalty(uint256 mintAmount) internal returns (uint256 totalRoyaltyMints) {
```
Comment (oveddan) [inline src/ERC721Drop.sol:1281]: can this have a better name that describes what it's doing? `_handle` is a bit vague.  maybe `_mintSupplyRoyalty` ?

File: src/ERC721Drop.sol [hunk 2]
```diff
@@ -1267,6 +1272,38 @@ contract ERC721Drop is
...
+            // If we have no recipient set, return 0 supply royalty.
+            if (royaltyRecipient == address(0)) {
+                return 0;
+            }
+            _mintNFTs(royaltyRecipient, totalRoyaltyMints);
+        }
+    }
+
+    function updateRoyaltyMintSchedule(uint32 newSchedule) external onlyAdmin {
+        if (newSchedule == 1) {
```
Comment (oveddan) [inline src/ERC721Drop.sol:1301]: why this error? What's the `1` represent?

File: src/ERC721Drop.sol [hunk 3]
```diff
@@ -1267,6 +1272,38 @@ contract ERC721Drop is
...
+        }
+    }
+
+    function _handleSupplyRoyalty(uint256 mintAmount) internal returns (uint256 totalRoyaltyMints) {
+        if (royaltyMintSchedule == 0) {
+            // If we still have no schedule, return 0 supply royalty.
+            return 0;
+        }
+
+        totalRoyaltyMints = (mintAmount + (_totalMinted() % royaltyMintSchedule)) / (royaltyMintSchedule - 1);
```
Comment (oveddan) [inline src/ERC721Drop.sol:1287]: can you add some comments explaining what this is doing?

File: src/ERC721Drop.sol [hunk 4]
```diff
@@ -1267,6 +1272,38 @@ contract ERC721Drop is
...
+
+        if (totalRoyaltyMints > 0) {
+            address royaltyRecipient = config.fundsRecipient;
+
+            // If we have no recipient set, return 0 supply royalty.
+            if (royaltyRecipient == address(0)) {
+                return 0;
+            }
+            _mintNFTs(royaltyRecipient, totalRoyaltyMints);
```
Comment (oveddan) [inline src/ERC721Drop.sol:1296]: should this be returning something since every other path returns 0 ?

File: src/ERC721Drop.sol (hunk 1 above)
Comment (oveddan) [reply src/ERC721Drop.sol:1281]: `mintQuantity` would be better - amount to be sounds like an eth value.

File: src/ERC721Drop.sol [hunk 5]
```diff
@@ -1267,6 +1272,38 @@ contract ERC721Drop is
...
+            return 0;
+        }
+
+        totalRoyaltyMints = (mintAmount + (_totalMinted() % royaltyMintSchedule)) / (royaltyMintSchedule - 1);
+
+        if (totalRoyaltyMints > 0) {
+            address royaltyRecipient = config.fundsRecipient;
+
+            // If we have no recipient set, return 0 supply royalty.
+            if (royaltyRecipient == address(0)) {
```
Comment (oveddan) [inline src/ERC721Drop.sol:1293]: how about moving this check earlier to avoid unnecessary computations above?

File: test/ERC721Drop.t.sol [hunk 6]
```diff
@@ -977,6 +977,49 @@ contract ERC721DropTest is Test {
         assertEq(dummyRenderer.someState(), "");
     }
 
+    function test_SupplyRoyaltyPurchase() public setupZoraNFTBase(100) {
+        vm.startPrank(DEFAULT_OWNER_ADDRESS);
+        vm.expectRevert(IERC721Drop.InvalidMintSchedule.selector);
+        zoraNFTBase.updateRoyaltyMintSchedule(1);
+
+        zoraNFTBase.updateRoyaltyMintSchedule(5);
```
Comment (oveddan) [inline test/ERC721Drop.t.sol:985]: what are these `1` and `5` values?  maybe they should come from constants with clear names

File: test/ERC721Drop.t.sol [hunk 7]
```diff
@@ -977,6 +977,49 @@ contract ERC721DropTest is Test {
         assertEq(dummyRenderer.someState(), "");
     }
 
+    function test_SupplyRoyaltyPurchase() public setupZoraNFTBase(100) {
```
Comment (oveddan) [inline test/ERC721Drop.t.sol:980]: this test is a bit hard to follow - and its one big test testing a bunch of logic - how about breaking it into smaller ones testing the different branches?

File: test/ERC721Drop.t.sol [hunk 8]
```diff
@@ -977,6 +977,49 @@ contract ERC721DropTest is Test {
...
+            presaleStart: 0,
+            presaleEnd: 0,
+            publicSalePrice: 0.1 ether,
+            maxSalePurchasePerAddress: 100,
+            presaleMerkleRoot: bytes32(0)
+        });
+        vm.stopPrank();
+
+        (, uint256 zoraFee) = zoraNFTBase.zoraFeeForAmount(9);
+        uint256 paymentAmount = 0.9 ether + zoraFee;
```
Comment (oveddan) [inline test/ERC721Drop.t.sol:999]: what's the 0.9 ether coming from?

File: test/ERC721Drop.t.sol [hunk 9]
```diff
@@ -977,6 +977,49 @@ contract ERC721DropTest is Test {
...
+        paymentAmount = 0.1 ether * 72 + zoraFee;
+        vm.expectRevert(IERC721Drop.Mint_SoldOut.selector);
+        zoraNFTBase.purchase{value: paymentAmount}(72);
+
+        (, zoraFee) = zoraNFTBase.zoraFeeForAmount(71);
+        paymentAmount = 0.1 ether * 71 + zoraFee;
+        zoraNFTBase.purchase{value: paymentAmount}(71);
+
+        assertEq(zoraNFTBase.balanceOf(address(456)), 80);
+        assertEq(zoraNFTBase.balanceOf(DEFAULT_FUNDS_RECIPIENT_ADDRESS), 20);
```
Comment (oveddan) [inline test/ERC721Drop.t.sol:1018]: one set of value assertions for all the possible branches is a bit brittle.  It would be better to have assertions for each possible branch/edge case.

**Comment by iainnash** [comment]
Left some comments – better testing would be good for this contract addition.

File: src/ERC721Drop.sol (hunk 4 above)
Comment (jgeary) [reply src/ERC721Drop.sol:1296]: the function returns `totalRoyaltyMints` which is set above

File: src/ERC721Drop.sol (hunk 2 above)
Comment (jgeary) [reply src/ERC721Drop.sol:1301]: this would mean every single mint goes to creator

Comment (oveddan) [reply src/ERC721Drop.sol:1301]: not quite sure I understand what the `1` means.  could you make the `1` an enum, or have it be a constant with a name explaining what `1` represents?

File: test/ERC721Drop.t.sol (hunk 9 above)
Comment (oveddan) [reply test/ERC721Drop.t.sol:1018]: also a fuzzy test would be great too - in particular around the royalty computation

File: src/ERC721Drop.sol (hunk 4 above)
Comment (oveddan) [reply src/ERC721Drop.sol:1296]: Oh I see - totalRoyaltyMints is assigned above. It looks like the return value is not being used by the calling function.   How about not returning anything then out of this function since the return value is not used?  Would make the code more readable and save a bit on gas.

File: src/ERC721Drop.sol (hunk 2 above)
Comment (oveddan) [reply src/ERC721Drop.sol:1301]: ok i get it now - schedule is how many mints would result in a royalty mint.  I'd add a comment here explaining why this is not allowed

File: src/ERC721Drop.sol [hunk 10]
```diff
@@ -1267,6 +1272,38 @@ contract ERC721Drop is
...
+                return 0;
+            }
+            _mintNFTs(royaltyRecipient, totalRoyaltyMints);
+        }
+    }
+
+    function updateRoyaltyMintSchedule(uint32 newSchedule) external onlyAdmin {
```
Comment (oveddan) [inline src/ERC721Drop.sol:1300]: There would be some issues that happen if this is updated after a few tokens have been minted already:

For example:
* the royalty mint schedule is 5
* 4 tokens are minted
* the royalty schedule is changed to 4
* a royalty would only be received when the mint count is 8

We should either:
* Not allow this to be updated after the first token is minted
* Store how many royalty mints there have been done so far, and compare that to how many there should be at this point, and mint the difference

Comment (jgeary) [reply src/ERC721Drop.sol:1300]: great point @oveddan , main goal was to get parity with 1155 which has this same impl but lets talk about this today

Comment (iainnash) [reply src/ERC721Drop.sol:1300]: @oveddan I think missing during update is expected – it's every Nth from that point forward

File: src/ERC721Drop.sol [hunk 11]
```diff
@@ -1267,6 +1275,36 @@ contract ERC721Drop is
...
+            return;
+        }
+
+        address royaltyRecipient = config.fundsRecipient;            
+        if (royaltyRecipient == address(0)) {
+            return;
+        }
+
+        uint256 totalRoyaltyMints = (mintQuantity + (_totalMinted() % royaltyMintSchedule)) / (royaltyMintSchedule - 1);
+        totalRoyaltyMints = MathUpgradeable.min(totalRoyaltyMints, config.editionSize - (mintQuantity + _totalMinted()));
```
Comment (iainnash) [inline src/ERC721Drop.sol:1295]: Should we have comments for this?

**Review by iainnash — APPROVED** [review]

File: src/ERC721Drop.sol (hunk 11 above)
Comment (jgeary) [reply src/ERC721Drop.sol:1295]: its @tbtstl 's whiteboard magic, would be great if you had any notes on it tyson.

**Review by kulkarohan — APPROVED** [review]
